go-controller/pkg/node/controllers/egressip/pod.go

```diff
@@ -0,0 +1,279 @@
+}
+
+func newPodIPConfigList() *podIPConfigList {
+	return &podIPConfigList{elems: []*podIPConfig{}}
+}
+
+func (pICL *podIPConfigList) Len() int {
+	return len(pICL.elems)
+}
+
+func (pICL *podIPConfigList) hasWithoutError(pIC2 *podIPConfig) bool {
+	for _, pIC := range pICL.elems {
+		if pIC.equal(pIC2) && !pIC.failed {
+			return true
+		}
+	}
+	return false
+}
+
+func (pICL *podIPConfigList) has(pIC2 *podIPConfig) bool {
+	for _, pIC := range pICL.elems {
+		if pIC.equal(pIC2) {
+			return true
+		}
+	}
+	return false
+}
+
+func (pICL *podIPConfigList) remove(idxs ...int) {
+	if len(idxs) == 0 {
+		return
+	}
+	newElems := make([]*podIPConfig, 0, len(pICL.elems))
+	idxToDelete := sets.New[int](idxs...)
+
+	for existingIdx, pIC := range pICL.elems {
+		if !idxToDelete.Has(existingIdx) {
+			newElems = append(newElems, pIC)
+		}
+	}
+	pICL.elems = newElems
+}
+
+// InsertOverwriteFailed should be used to add elements to the podIPConfigList that have failed to be applied.
+func (pICL *podIPConfigList) InsertOverwriteFailed(pICs ...podIPConfig) {
+	for _, pIC := range pICs {
+		pIC.failed = true
+		pICL.insertOverwrite(pIC)
+	}
+}
+
+// Insert should be used to add elements to the podIPConfigList
+func (pICL *podIPConfigList) Insert(pICs ...podIPConfig) {
+	for _, pc := range pICs {
+		pICL.insertOverwrite(pc)
+	}
+}
+
+func (pICL *podIPConfigList) insertOverwrite(pIC podIPConfig) {
+	var dupeIdxs []int
+	for idx, existingPIC := range pICL.elems {
+		if existingPIC.equal(&pIC) {
+			dupeIdxs = append(dupeIdxs, idx)
+		}
+	}
+	pICL.remove(dupeIdxs...)
+	pICL.elems = append(pICL.elems, &pIC)
+}
+
+func (c *Controller) onPodAdd(obj interface{}) {
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", &corev1.Pod{}, obj))
+		return
+	}
+	if pod == nil {
+		utilruntime.HandleError(errors.New("invalid Pod provided to onPodAdd()"))
+		return
+	}
+	// if the pod does not have IPs AND there are no multus network status annotations found, skip it
+	if len(pod.Status.PodIPs) == 0 && len(pod.Annotations[nettypes.NetworkStatusAnnot]) == 0 {
+		return
+	}
+	c.podQueue.Add(pod)
+}
+
+func (c *Controller) onPodUpdate(oldObj, newObj interface{}) {
+	o, ok := oldObj.(*corev1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", &corev1.Pod{}, o))
+		return
+	}
+	n, ok := newObj.(*corev1.Pod)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expecting %T but received %T", &corev1.Pod{}, n))
+		return
+	}
+	if o == nil || n == nil {
+		utilruntime.HandleError(errors.New("invalid Pod provided to onPodUpdate()"))
+		return
+	}
+	// if labels AND assigned Pod IPs AND the multus network status annotations are the same, skip processing changes to the pod.
+	if reflect.DeepEqual(o.Labels, n.Labels) &&
+		reflect.DeepEqual(o.Status.PodIPs, n.Status.PodIPs) &&
+		reflect.DeepEqual(o.Annotations[nettypes.NetworkStatusAnnot], n.Annotations[nettypes.NetworkStatusAnnot]) {
```

Comment:
hmm why do we care about the network status annotation?

Reply:
We dont - removed.